web3/utils/validation.py

```diff
@@ -27,6 +33,26 @@
     length_of_array_type,
     sub_type_of_array_type,
 )
+from web3.utils.formatters import (
+    apply_formatter_to_array,
+)
+from web3.utils.toolz import (
+    compose,
+    groupby,
+    valfilter,
+    valmap,
+)
+
+
+def _prepare_selector_collision_msg(duplicates):
+    dup_sel = valmap(apply_formatter_to_array(abi_to_signature), duplicates)
+    joined_func_groups = map(lambda funcs: ', '.join(funcs), dup_sel.values())
+    sel_func_groups = zip(joined_func_groups, dup_sel.keys())
```

Comment:
I don't think `dup_sel`'s keys and values are guaranteed to be iterated in the same order in the previous two lines.

Some other options to consider:

1.
```py
def _duplicate_selector_messages(duplicate_selectors):
    for selector, functions in duplicate_selectors.items():
        yield ', '.join(functions) + ' have selector ' + selector

def _prepare_selector_collision_msg(duplicates):
    dup_sel = valmap(apply_formatter_to_array(abi_to_signature), duplicates)
    func_sel_msg_list = list(_duplicate_selector_messages(dup_sel))
```

2.
```py
def _prepare_selector_collision_msg(duplicates):
    dup_sel = valmap(apply_formatter_to_array(abi_to_signature), duplicates)
    joined_funcs = valmap(lambda funcs: ', '.join(funcs), dup_sel)
    func_sel_msg_list = [funcs + ' have selector ' + sel for sel, funcs in joined_funcs.items()]
```

(all untested)